File: JasSwift/Sources/Menu/FifoActionRouting.swift
import Foundation

// Native-first routing for the test-only `--test-fifo` `action <name>`
// command channel. JasSwift port of the Python reference
// (jas_app.py MainWindow._dispatch_action_by_name, commit eae9c2f9).
//
// WHY native-first: document-mutating menubar / edit / file actions
// (`new_document`, `select_all`, `delete_selection`, ...) are
// NATIVE-INTERCEPTED. Their actions.yaml `effects` are deliberate
// `log` / `if` STUBS — the real behavior lives natively (JasCommands'
// private menu router + keyboard-only natives). The FIFO `action` verb
// used to route every name through the GENERIC panel dispatcher
// (`LayersPanel.dispatchYamlAction`), which runs only those `effects`,
// so `action select_all` / `delete_selection` logged-and-no-op'd while a
// real menu click / keystroke mutated the document. `tool <id>` is fine
// because select_tool's real behavior IS its generic `set: active_tool`
// effect, so it does NOT go through here.
//
// JasCommands.dispatchMenuAction is a PRIVATE method on a SwiftUI
// `struct JasCommands: Commands` (@FocusedValue model + injected
// closures) and is NOT reachable from the app delegate's FIFO handler.
// So this dispatcher performs the SAME document-mutation ops the
// JasCommands handlers use, directly on the active model:
//   select_all       -> Controller.selectAll()      (== JasCommands.selectAll)
//   delete_selection -> shared opApply delete_selection in a named txn
//                       (== the keyboard-only native delete; one named
//                        undo step via the shared op_apply dispatcher,
//                        matching Python's _route_delete_selection)
// Everything else falls through to the generic panel dispatcher for
// genuine panel / generic-effect actions (panel toggles, etc.).
public enum FifoActionRouting {
    /// Native-first dispatch of a named action for the FIFO `action`
    /// channel. Routes document-mutating natives directly on `model`,
    /// else calls `fallthrough` (the generic panel dispatcher).
    ///
    /// `fallthrough` is injectable so the routing decision is unit-
    /// testable without standing up the YAML effects pipeline (the Swift
    /// analog of Python monkeypatching dock_panel._dispatch_yaml_action).
    /// It defaults to `LayersPanel.dispatchYamlAction`, exactly what the
    /// FIFO handler used before this native-first layer.
    public static func dispatch(
        _ name: String,
        model: Model,
        params: [String: Any] = [:],
        fallthrough fallthroughDispatch:
            (_ name: String, _ model: Model, _ params: [String: Any]) -> Void
            = { name, model, params in
                LayersPanel.dispatchYamlAction(name, model: model, params: params)
            }
    ) {
        switch name {
        case "select_all":
            // Same op JasCommands.selectAll() runs (selection-only,
            // non-undoable per OP_LOG.md §7/§8).
            Controller(model: model).selectAll()
        case "delete_selection":
            // Keyboard-only native delete: journals ONE named
            // `delete_selection` op via the shared opApply dispatcher
            // (OP_LOG.md §9), matching Python's _route_delete_selection.
            // No-op when nothing is selected (opApply handles the empty
            // case). The orphan-confirm NSAlert that JasCommands shows is
            // intentionally skipped here: the FIFO channel is headless
            // test-only input with no UI to confirm against.
            guard !model.document.selection.isEmpty else { return }
            model.withTxn {
                model.nameTxn("delete_selection")
                opApply(model, Controller(model: model), ["op": "delete_selection"])
            }
        default:
            // Genuine panel / generic-effect action.
            fallthroughDispatch(name, model, params)
        }
    }
}

File: jas_ocaml/lib/panels/fifo_action_routing.ml
(* Native-first routing for named workspace actions reaching the app from a
   name-driven caller (today: the test-only [--test-fifo] [action <name>]
   channel in bin/main.ml).

   WHY native-first: document-mutating menubar / edit actions ([select_all],
   [delete_selection], ...) are NATIVE-INTERCEPTED. Their actions.yaml
   [effects] are deliberate [log] / [if] STUBS — the real behavior lives in
   native OCaml code (the menubar [on_menu_action] closure and the keyboard
   delete path in bin/main.ml). So routing such a name through the GENERIC
   dispatcher ([Panel_menu.dispatch_yaml_action], which runs only an action
   YAML [effects]) logs-and-no-ops, while a real menu click / keystroke
   mutates the document. The FIFO [action] verb formerly went straight to the
   generic dispatcher and therefore silently dropped these mutations.

   This dispatcher routes the native-intercepted document mutations through
   the SAME native ops the menu / keyboard handlers use, then falls through
   to the generic dispatcher for genuine panel / generic-effect actions
   (e.g. [select_tool] panel toggles), whose real behavior IS their YAML
   effect. Mirrors the Python MainWindow._dispatch_action_by_name
   (jas/jas_app.py) for cross-app equivalence. *)

(* select_all: the SAME op the menubar Edit > Select All arm runs
   (menubar.ml [on_menu_action] [select_all] arm). Selection-only, not a
   journaled mutation (Controller#select_all writes via
   set_document_unbracketed). *)
let select_all (model : Model.model) : unit =
  (new Controller.controller ~model ())#select_all

(* delete_selection: route through the SHARED [Op_apply.op_apply] dispatcher
   exactly like the keyboard Delete path in bin/main.ml — a named transaction
   wrapping one [delete_selection] op, so the gesture journals ONE named undo
   step (the same [Document.delete_selection] body). The txn name matches the
   keyboard path ([delete_orphan_confirm_ok]) so a FIFO delete and a keyboard
   delete are byte-identical in the journal.

   Headless: the keyboard path also runs a GUI orphan-confirm BEFORE this
   mutation (a modal that needs a parent window). That confirm is a UI concern
   and is intentionally NOT replicated here — this dispatcher is a pure,
   testable, window-free seam. On a no-orphan selection the two paths are
   equivalent; an orphan-confirm gate, if ever wanted on the FIFO path, would
   wrap the call site in bin/main.ml, not this function. *)
let delete_selection (model : Model.model) : unit =
  let doc = model#document in
  if Document.PathMap.is_empty doc.Document.selection then ()
  else begin
    let ctrl = new Controller.controller ~model () in
    model#with_txn (fun () ->
      model#name_txn "delete_orphan_confirm_ok";
      Op_apply.op_apply model ctrl
        (`Assoc [ ("op", `String "delete_selection") ]))
  end

let dispatch
    ?(fallthrough :
        (params:(string * Yojson.Safe.t) list -> string -> Model.model -> unit) option)
    ?(params : (string * Yojson.Safe.t) list = [])
    (name : string) (model : Model.model) : unit =
  (* Default fall-through is the generic panel dispatcher. Injectable so a
     unit test can spy on it (mirrors the Python spy/replace of
     dock_panel._dispatch_yaml_action). *)
  let fallthrough =
    match fallthrough with
    | Some f -> f
    | None -> (fun ~params name model ->
        Panel_menu.dispatch_yaml_action ~params name model)
  in
  match name with
  | "select_all" -> select_all model
  | "delete_selection" -> delete_selection model
  | _ -> fallthrough ~params name model
